File: 1582-design-browser-history/1582-design-browser-history.py
class ListNode:
    def   __init__(self, val = None, next = None, prev = None):
        self.val = val
        self.next = next
        self.prev = prev


class BrowserHistory(object):

    def __init__(self, homepage):
        """
        :type homepage: str
        """
        self.curr = ListNode(homepage)
        

    def visit(self, url):
        """
        :type url: str
        :rtype: None
        """
        newUrl = ListNode(url, None, self.curr)

        self.curr.next = newUrl
        self.curr = self.curr.next

    def back(self, steps):
        """
        :type steps: int
        :rtype: str
        """
        # returnMe = None
        while self.curr.prev and steps > 0:
            self.curr = self.curr.prev
            steps -= 1
        return self.curr.val
        

    def forward(self, steps):
        """
        :type steps: int
        :rtype: str
        """
        while self.curr.next and steps > 0:
            self.curr = self.curr.next
            steps -= 1
        return  self.curr.val
        


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)

File: 1582-design-browser-history/1582-design-browser-history.js
/**
 * @param {string} homepage
 */

var Node = function(value) {
    this.val = value;
    this.next = null;
    this.pre = null;
};


var BrowserHistory = function(homepage) {
    this.head = new Node(homepage);
    this.tail = this.head;
    this.curr = this.head;
    this.length = 1; 
};

/** 
 * @param {string} url
 * @return {void}
 */
BrowserHistory.prototype.visit = function(url) {
    if(this.curr === this.tail){
        this.curr.next = new Node(url);
        this.curr.next.pre = this.curr;
        this.curr =this.curr.next;
        this.tail = this.curr;
    }else{
        this.curr.next = new Node(url);
        this.curr.next.pre = this.curr;
        this.curr = this.curr.next;
        this.tail = this.curr;
    }
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.back = function(steps) {
    
    while(steps !== 0 && this.curr.pre !== null){
        this.curr = this.curr.pre;
        steps -= 1;
    }

    return this.curr.val;
};

/** 
 * @param {number} steps
 * @return {string}
 */
BrowserHistory.prototype.forward = function(steps) {
    
    while(steps !== 0 && this.curr.next !== null){
        this.curr = this.curr.next;
        steps -= 1;
    }

    return this.curr.val;
};

/** 
 * Your BrowserHistory object will be instantiated and called as such:
 * var obj = new BrowserHistory(homepage)
 * obj.visit(url)
 * var param_2 = obj.back(steps)
 * var param_3 = obj.forward(steps)
 */